python/oauth-demo/app.py

```diff
@@ -0,0 +1,106 @@
+import os
+import uuid
+import requests
+from flask import Flask, request, redirect, jsonify, session, send_from_directory
+from dotenv import load_dotenv
+import asana
+from asana.rest import ApiException
+
+# Load environment variables from the .env file
+load_dotenv()
+
+app = Flask(__name__)
+# Set a secret key to securely sign session cookies and protect session data integrity
+app.secret_key = os.urandom(24)
+
+# Serve the main HTML page (index.html) from the 'static' directory.
+@app.route('/')
+def index():
+    return send_from_directory('static', 'index.html')
+
+@app.route('/authenticate')
+def authenticate():
+    # Generate a state value and store it in the session
+    state = str(uuid.uuid4())
+    session['state'] = state
+
+    # Redirect user to Asana's OAuth authorization endpoint
+    # Documentation: https://developers.asana.com/docs/oauth#user-authorization-endpoint
+    asana_url = (
+        f"https://app.asana.com/-/oauth_authorize?response_type=code&client_id={os.getenv('CLIENT_ID')}"
+        f"&redirect_uri={os.getenv('REDIRECT_URI')}&state={state}"
+    )
+    return redirect(asana_url)
+
+@app.route('/oauth-callback')
+def oauth_callback():
+    # Get the state from the query parameters and compare it with the one stored in the session
+    state = request.args.get('state')
+    if state != session.get('state'):
+        return "State mismatch. Possible CSRF attack.", 422
+
+    # Get the code from the callback query
+    code = request.args.get('code')
+
+    # Payload for the token exchange
+    # Documentation: https://developers.asana.com/docs/oauth#token-exchange-endpoint
+    token_url = 'https://app.asana.com/-/oauth_token'
+    data = {
+        'grant_type': 'authorization_code',
+        'client_id': os.getenv('CLIENT_ID'),
```

Comment:
Small, improvement is adding a function to check that the environment variables are missing. Also, might be more clear to add those variables at the top instead of calling `os.getenv()` where used. based on the `.env` file having placeholder values I don't think this is needed but just a suggestion

```
def get_env_var(name):
    value = os.getenv(name)
    if value is None:
        raise RuntimeError(f"The environment variable '{name}' is not set.")
    return value

CLIENT_ID = get_env_var('CLIENT_ID')
CLIENT_SECRET = get_env_var('CLIENT_SECRET')
REDIRECT_URI = get_env_var('REDIRECT_URI')
```